x/mongo/driver/driver.go

```diff
@@ -24,6 +25,65 @@ import (
 	"go.mongodb.org/mongo-driver/x/mongo/driver/session"
 )
 
+// AuthConfig holds the information necessary to perform an authentication attempt.
+// this was moved from the auth package to avoid a circular dependency. The auth package
+// reexports this under the old name to avoid breaking the public api.
+type AuthConfig struct {
+	Description   description.Server
+	Connection    Connection
+	ClusterClock  *session.ClusterClock
+	HandshakeInfo HandshakeInformation
+	ServerAPI     *ServerAPIOptions
+	HTTPClient    *http.Client
+}
+
+// OIDCCallback is the type for both Human and Machine Callback flows. RefreshToken will always be
+// nil in the OIDCArgs for the Machine flow.
+type OIDCCallback func(context.Context, *OIDCArgs) (*OIDCCredential, error)
+
+// OIDCArgs contains the arguments for the OIDC callback.
+type OIDCArgs struct {
+	Version      int
+	Timeout      time.Time
+	IDPInfo      *IDPInfo
+	RefreshToken *string
+}
+
+// OIDCCredential contains the access token and refresh token.
+type OIDCCredential struct {
+	AccessToken  string
+	ExpiresAt    *time.Time
+	RefreshToken *string
+}
+
+// IDPInfo contains the information needed to perform OIDC authentication with an Identity Provider.
+type IDPInfo struct {
+	Issuer        string   `bson:"issuer"`
+	ClientID      string   `bson:"clientId"`
+	RequestScopes []string `bson:"requestScopes"`
+}
+
```

Comment:
I think this is my biggest question on design right here. I could add another interface called `Reauthenticator` and downcast in operation.go on a 391 error to `Reauthenticator` so that every auth mechanism doesn't need to have a definition for `Reauth`, but I like this approach since it will require any new authentication mechanism to consider whether or not it supports Reauth than simply forgetting to do that, since this moves the check to compile time instead of runtime. This is the rust programmer in me, so I'll defer to whatever others prefer

Reply:
I agree, and the server may also decide to add Reauth capability to existing mechanisms.

Reply:
I most often see "upgrading" Go interfaces used as a backward compatibility mechanism. IMO some extra no-op functions is better than runtime type assertions. I agree with the current approach.